vllm/engine/llm_engine.py

```diff
@@ -1320,15 +1343,19 @@ def _process_model_outputs(self, virtual_engine: int,
                 self.output_processor.process_outputs(seq_group, output,
                                                       is_async)
 
-        # Free the finished sequence groups.
+        # For async + multi-step, free finished seqs and create outputs
+        # only on the final step.
+        if is_multi_step and not is_last_output:
+            return
+
         for scheduler in self.scheduler:
             scheduler.free_finished_seq_groups()
 
         # Create the outputs.
         for i, _ in enumerate(seq_group_metadata_list):
             scheduled_seq_group = scheduler_outputs.scheduled_seq_groups[i]
 
-            if i in finished_before:
+            if not is_multi_step and i in finished_before:
```

Comment:
why add `not is_multi_step` here? shouldn't double processing also be avoided for multi-step?

Reply:
Actually not since multi-step is using is_last_output to indicate if to run this code or not.